src/Database/Kioku/Internal/TrieIndex.hs

```diff
@@ -23,22 +24,34 @@ import            System.IO
 import            Database.Kioku.Internal.Buffer
 import            Database.Kioku.Memorizable
 
-
 trieLookup :: BS.ByteString -> TrieIndex -> [Int]
-trieLookup key = maybe [] trieRootElems . lookupSubtrie key
+trieLookup key = maybe [] (trieRootElems []) . lookupSubtrie key
 
 trieMatch :: BS.ByteString -> TrieIndex -> [Int]
 trieMatch prefix = maybe [] trieElems . lookupSubtrie prefix
 
+trieAllHitsAlong :: BS.ByteString -> TrieIndex -> [Int]
+trieAllHitsAlong = go []
+  where
+    go results "" trie = trieRootElems results trie
+    go results key trie@(TI _ _ offset) =
+      let (p, ath) = BS.splitAt 1 key
+       in case lookupSubtrie p trie of
+            Nothing -> trieRootElems results trie
+            Just subtrie@(TI _ _ subOffset) ->
+              if offset == subOffset
```

Comment:
This is very subtle... what's going on here? Is this somehow guarding against double-returning results? If so, it's worth of at least a comment if not an actual test.

Reply:
This is to eliminate unnecessary calls to `trieRootElems` so that we don't deserialize tries needlessly. There's no reason to read the trie when we know we haven't matched the key (arcDrop == length key) and so we only call `trieRootElems` upon entering a new subtrie. We also call it in the Nothing or  "" key cases for when our key exactly matches.
It would be slightly more efficient to have the remaining arc of the trie in hand so we are not having the peek at it in each call to `lookupSubtrie` but I think would make the code quite a bit uglier.

Reply:
In that case, rather than have this query function pick apart the `TI` constructors, I would extract some function with a meaningful name to check whether the two tries are different in the way you're trying to detect. Give it a "logical" name that helps the reader understand what you're getting at (i.e. not simply `isSameOffset` or `isDifferentOffset`). Then I would still add a comment here with the `if` where that function would be used to explain.